proposals/NNNN-noncopyable-generics.md

```diff
@@ -0,0 +1,713 @@
+            - [Conforming to `Copyable`](#conforming-to-copyable)
+        - [Type parameters](#type-parameters)
+        - [Structs and enums](#structs-and-enums)
+        - [Classes](#classes)
+        - [Protocols](#protocols)
+            - [Protocol Extensions](#protocol-extensions)
+            - [Protocol Inheritance](#protocol-inheritance)
+    - [Detailed Design](#detailed-design)
+        - [Conditionally-copyable types](#conditionally-copyable-types)
+        - [The top type](#the-top-type)
+        - [Existentials](#existentials)
+        - [Diagnosing contradictory intent](#diagnosing-contradictory-intent)
+        - [Scoping rule](#scoping-rule)
+    - [Source Compatibility](#source-compatibility)
+    - [ABI compatibility](#abi-compatibility)
+    - [Implications on adoption](#implications-on-adoption)
+    - [Alternatives Considered](#alternatives-considered)
+        - [`NonCopyable` as a Positive Requirement](#noncopyable-as-a-positive-requirement)
+        - [`❌Copyable` as a Negative Requirement](#❌copyable-as-a-negative-requirement)
+        - [Alternative Spellings](#alternative-spellings)
+        - [Inferred Conditional Copyability](#inferred-conditional-copyability)
+        - [Extension defaults](#extension-defaults)
+    - [Future Directions](#future-directions)
+        - [Standard Library and Concurrency support](#standard-library-and-concurrency-support)
+        - [Tuples and element packs](#tuples-and-element-packs)
+        - [`~Escapable`](#escapable)
+        - [Non-copyable Classes](#non-copyable-classes)
+        - [Dynamic Queries and Runtime Support](#dynamic-queries-and-runtime-support)
+    - [Acknowledgments](#acknowledgments)
+
+<!-- markdown-toc end -->
+
+## Introduction
+
+The noncopyable types introduced in [SE-0390: Noncopyable structs and enums](0390-noncopyable-structs-and-enums.md) come with the heavy limitation that such values cannot be substituted for a generic type parameter, erased to an existential, or conform to a protocol.
+This proposal extends Swift's type system with syntax and semantics allowing noncopyable types to be used in all of these ways.
+
+## Motivation
+
+SE-0390 laid the groundwork for declaring struct and enum types that cannot be copied.
+This ensures correct semantics for types for which it is not meaningful to
+have multiple copies:
+```swift
+// A file descriptor cannot be usefully used from
+// multiple places at once, so make it non-copyable
+// to prevent such usage.
+struct FileDescriptor: ~Copyable { ... }
+```
+
+This also provides an alternative to class objects for some use cases.
+In particular, the tightly defined lifecycle allows noncopyable structs
+to carry deinitializers:
+```swift
+struct HeapBuffer: ~Copyable {
+  init() { ... allocate working storage on heap ... }
+  deinit() { ... release storage ... }
+}
+```
+
+But SE-0390 also made a number of concessions to simplify the initial implementation,
```

Comment:
I would re-frame this in terms of making incremental progress on supporting non-copyable types in the language. SE-0390 is a useful step on its own, and the generics interactions are complicated with a lot of nuanced design decisions (as demonstrated in this proposal!) so it was best to subset it out into its own proposal to carefully consider the implications of those interactions in the generics system.